include/irr/asset/filters/CBasicImageFilterCommon.h

```diff
@@ -0,0 +1,246 @@
+// Copyright (C) 2020- Mateusz 'DevSH' Kielan
+// This file is part of the "IrrlichtBAW" engine.
+// For conditions of distribution and use, see copyright notice in irrlicht.h
+
+#ifndef __IRR_C_BASIC_IMAGE_FILTER_COMMON_H_INCLUDED__
+#define __IRR_C_BASIC_IMAGE_FILTER_COMMON_H_INCLUDED__
+
+#include "irr/core/core.h"
+
+#include "irr/asset/IImageFilter.h"
+
+namespace irr
+{
+namespace asset
+{
+
+class CBasicImageFilterCommon
+{
+	public:
+		struct TexelBlockInfo
+		{
+			TexelBlockInfo(E_FORMAT format) :
+				dimension(getBlockDimensions(format)),
+				maxCoord(dimension-core::vector3du32_SIMD(1u,1u,1u))
+			{}
+
+			core::vector3du32_SIMD dimension;
+			core::vector3du32_SIMD maxCoord;
+		};
+		static inline auto texelsToBlocks(const core::vector3du32_SIMD& coord, const TexelBlockInfo& info)
+		{
+			return (coord+info.maxCoord)/info.dimension;
+		}
+
+		template<typename F>
+		static inline void executePerBlock(const ICPUImage* image, const IImage::SBufferCopy& region, F& f)
+		{
+			const auto& params = image->getCreationParameters();
+			const auto& extent = params.extent;
+
+			TexelBlockInfo blockInfo(params.format);
+
+			core::vector3du32_SIMD trueOffset;
+			trueOffset.x = region.imageOffset.x;
+			trueOffset.y = region.imageOffset.y;
+			trueOffset.z = region.imageOffset.z;
+			trueOffset = texelsToBlocks(trueOffset,blockInfo);
+			
+			core::vector3du32_SIMD trueExtent = texelsToBlocks(region.getTexelStrides(),blockInfo);
+
+			trueExtent.w = asset::getTexelOrBlockBytesize(params.format);
+			for (uint32_t layer=0u; layer<region.imageSubresource.layerCount; layer++)
+			for (uint32_t zBlock=trueOffset.z; zBlock<trueExtent.z; ++zBlock)
+			for (uint32_t yBlock=trueOffset.y; yBlock<trueExtent.y; ++yBlock)
+			for (uint32_t xBlock=trueOffset.x; xBlock<trueExtent.x; ++xBlock)
+			{
+				auto texelPtr = region.bufferOffset+((((region.imageSubresource.baseArrayLayer+layer)*trueExtent[2]+zBlock)*trueExtent[1]+yBlock)*trueExtent[0]+xBlock)*trueExtent[3];
```

Comment:
ahhh I remember the pain till now :D

Reply:
merge again, now its perfect ;)

Reply:
yep, I'm about to merge it - the GLI writer will be deeply tested:)